File: supervised/classification/knn/golang/knn.go
package knn

import (
	"fmt"
	"math"
	"sort"
)

// DistanceMetric enum type
type DistanceMetric string

const (
	Euclidean DistanceMetric = "euclidean"
	Manhattan DistanceMetric = "manhattan"
)

// WeightMethod enum type
type WeightMethod string

const (
	Uniform  WeightMethod = "uniform"
	Distance WeightMethod = "distance"
)

// KNNClassifier represents a K-Nearest Neighbors classifier
type KNNClassifier struct {
	k             int
	distanceMetric DistanceMetric
	weightMethod  WeightMethod
	xTrain        [][]float64
	yTrain        []float64
}

// NewKNNClassifier creates a new KNN classifier with the specified parameters
func NewKNNClassifier(k int, distanceMetric DistanceMetric, weightMethod WeightMethod) *KNNClassifier {
	return &KNNClassifier{
		k:             k,
		distanceMetric: distanceMetric,
		weightMethod:  weightMethod,
		xTrain:        [][]float64{},
		yTrain:        []float64{},
	}
}

// Fit trains the KNN model with the provided training data
func (knn *KNNClassifier) Fit(xTrain [][]float64, yTrain []float64) error {
	if len(xTrain) != len(yTrain) {
		return fmt.Errorf("data and labels must have the same length")
	}
	
	// Make deep copies of the data to avoid reference issues
	knn.xTrain = make([][]float64, len(xTrain))
	for i, x := range xTrain {
		knn.xTrain[i] = make([]float64, len(x))
		copy(knn.xTrain[i], x)
	}
	
	knn.yTrain = make([]float64, len(yTrain))
	copy(knn.yTrain, yTrain)
	
	return nil
}

// Calculate distance between two vectors
func (knn *KNNClassifier) distance(a, b []float64) (float64, error) {
	if len(a) != len(b) {
		return 0, fmt.Errorf("vectors must have the same length")
	}
	
	switch knn.distanceMetric {
	case Euclidean:
		sum := 0.0
		for i := 0; i < len(a); i++ {
			sum += math.Pow(a[i]-b[i], 2)
		}
		return math.Sqrt(sum), nil
		
	case Manhattan:
		sum := 0.0
		for i := 0; i < len(a); i++ {
			sum += math.Abs(a[i] - b[i])
		}
		return sum, nil
		
	default:
		return 0, fmt.Errorf("unsupported distance metric: use 'euclidean' or 'manhattan'")
	}
}

// Neighbor represents a data point and its distance from a query point
type Neighbor struct {
	Label    float64
	Distance float64
}

// PredictSingle predicts the label for a single data point
func (knn *KNNClassifier) PredictSingle(x []float64) (float64, error) {
	if len(knn.xTrain) == 0 {
		return 0, fmt.Errorf("model has not been trained yet")
	}

	// Calculate distances to all training points
	neighbors := make([]Neighbor, len(knn.xTrain))
	for i, xTrain := range knn.xTrain {
		dist, err := knn.distance(x, xTrain)
		if err != nil {
			return 0, err
		}
		
		neighbors[i] = Neighbor{
			Label:    knn.yTrain[i],
			Distance: dist,
		}
	}
	
	// Sort by distance
	sort.Slice(neighbors, func(i, j int) bool {
		return neighbors[i].Distance < neighbors[j].Distance
	})
	
	// Get k-nearest neighbors
	k := knn.k
	if k > len(neighbors) {
		k = len(neighbors)
	}
	
	kNearest := neighbors[:k]
	
	// Determine the predicted label based on weight method
	switch knn.weightMethod {
	case Uniform:
		// Count labels
		counts := make(map[float64]int)
		for _, neighbor := range kNearest {
			counts[neighbor.Label]++
		}
		
		// Find the most frequent label
		var maxCount int
		var predictedLabel float64
		for label, count := range counts {
			if count > maxCount {
				maxCount = count
				predictedLabel = label
			}
		}
		
		return predictedLabel, nil
		
	case Distance:
		// Weighted voting based on distance
		weights := make(map[float64]float64)
		for _, neighbor := range kNearest {
			// Add small epsilon to avoid division by zero
			weight := 1.0 / (neighbor.Distance + 1e-5)
			weights[neighbor.Label] += weight
		}
		
		// Find the label with the highest weighted vote
		var maxWeight float64
		var predictedLabel float64
		for label, weight := range weights {
			if weight > maxWeight {
				maxWeight = weight
				predictedLabel = label
			}
		}
		
		return predictedLabel, nil
		
	default:
		return 0, fmt.Errorf("unsupported weight method: use 'uniform' or 'distance'")
	}
}

// Predict predicts labels for multiple data points
func (knn *KNNClassifier) Predict(X [][]float64) ([]float64, error) {
	if len(knn.xTrain) == 0 {
		return nil, fmt.Errorf("model has not been trained yet")
	}
	
	predictions := make([]float64, len(X))
	
	for i, x := range X {
		prediction, err := knn.PredictSingle(x)
		if err != nil {
			return nil, err
		}
		predictions[i] = prediction
	}
	
	return predictions, nil
}

// Score calculates the accuracy of predictions compared to true labels
func (knn *KNNClassifier) Score(X [][]float64, yTrue []float64) (float64, error) {
	if len(X) != len(yTrue) {
		return 0, fmt.Errorf("data and labels must have the same length")
	}
	
	yPred, err := knn.Predict(X)
	if err != nil {
		return 0, err
	}
	
	correct := 0
	for i := 0; i < len(yTrue); i++ {
		if yPred[i] == yTrue[i] {
			correct++
		}
	}
	
	return float64(correct) / float64(len(yTrue)), nil
} 

File: supervised/classification/knn/js/knn.js
const fs = require('fs'); // for reading csv

class KNNClassifier {
    constructor(k=3, distanceMetric='euclidean', weights='uniform') {
        this.k = k; // Number of nearest neighbors
        this.distanceMetric = distanceMetric; // Distance metric
        this.weights = weights; // Weighting method
        this.X_train = [];
        this.y_train = [];
    }

    // Add training data
    fit(X_train, y_train) {
        if (X_train.length !== y_train.length) {
            throw new Error("Data and labels must have the same length.");
        }
        this.X_train = X_train;
        this.y_train = y_train;
    }

    // Distance calculation
    _distance(a, b) {
        if (this.distanceMetric === 'euclidean') {
            return Math.sqrt(a.reduce((sum, val, i) => sum + (val - b[i]) ** 2, 0));
        } else if (this.distanceMetric === 'manhattan') {
            return a.reduce((sum, val, i) => sum + Math.abs(val - b[i]), 0);
        } else {
            throw new Error("Unsupported distance metric: use 'euclidean' or 'manhattan'");
        }
    }


    // Predict the label for a single data point
    predictSingle(x) {
        const distances = this.X_train.map((x_train, i) => ({
            label: this.y_train[i],
            distance: this._distance(x, x_train),
          }));
      
          // Sort by distance
          distances.sort((a, b) => a.distance - b.distance);
          const neighbors = distances.slice(0, this.k);
      
          if (this.weights === 'uniform') {
            // Count labels
            const counts = {};
            for (const neighbor of neighbors) {
              counts[neighbor.label] = (counts[neighbor.label] || 0) + 1;
            }
            // Return the most frequent
            return Object.entries(counts).reduce((a, b) => (a[1] > b[1] ? a : b))[0];
          }
      
          else if (this.weights === 'distance') {
            const weights = {};
            for (const neighbor of neighbors) {
              const weight = 1 / (neighbor.distance + 1e-5); // avoid div by 0
              weights[neighbor.label] = (weights[neighbor.label] || 0) + weight;
            }
            return Object.entries(weights).reduce((a, b) => (a[1] > b[1] ? a : b))[0];
          }
      
          else {
            throw new Error("Unsupported weight method: use 'uniform' or 'distance'");
          }
    }

    // Predict labels for multiple data points
    predict(X) {
        if (!this.X_train.length) {
            throw new Error("Model has not been trained yet.");
        }

        if (!Array.isArray(X)) {
            throw new Error("Input data must be an array.");
        }

        if (X.length === 0) {
            return [];
        }

        return X.map((point) => this.predictSingle(point));
    }

    score(X, y_true) {
        const y_pred = this.predict(X);
        let correct = 0;
        for (let i = 0; i < y_true.length; i++) {
          if (parseFloat(y_pred[i]) === parseFloat(y_true[i])) correct++;
        }
        return correct / y_true.length;
    }
}


function mulberry32(seed) {
    return function () {
        seed |= 0;
        seed = seed + 0x6D2B79F5 | 0;
        let t = Math.imul(seed ^ seed >>> 15, 1 | seed);
        t = t + Math.imul(t ^ t >>> 7, 61 | t) ^ t;
        return ((t ^ t >>> 14) >>> 0) / 4294967296;
    }
}

function shuffle(array, randomFn) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(randomFn() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
}


// Load data from CSV
// make sure to run this from the js/ folder
const filePath = `../../../../dataset/customer_segmentation/clean/data.csv`;
const data = fs.readFileSync(filePath, 'utf8');
let rows = data.trim().split('\n').map(row => row.split(','));

console.log(`Total rows: ${rows.length}`);
console.log(`Total columns: ${rows[0].length}`);
console.log(`Columns`, rows[0])

// Remove the header row
rows = rows.slice(1);

// Feature selection
// Drop 'Work_Experience', 'Gender_Male', 'Gender_Female', 'Family_Size'
rows = rows.map(row => {
    return [
        parseFloat(row[0]), // Age
        row[5] === `True`? 1 : 0, // Ever_Married_No
        row[6] === `True`? 1 : 0, // Ever_Married_Yes
        row[7] === `True`? 1 : 0, // Graduated_No
        row[8] === `True`? 1 : 0, // Graduated_Yes
        parseFloat(row[9]), // Profession_Freq
        parseFloat(row[10]), // Spending_Score_Freq,
        parseFloat(row[11]), // Category_Freq
        parseFloat(row[12]), // Segmentation_Freq
    ]
})

// Shuffle data
// Set seed for reproducibility
const seed = 1234;
const rand = mulberry32(seed);
shuffle(rows, rand);

// Split data into training and test sets
const splitIndex = Math.floor(rows.length * 0.8);
const trainingData = rows.slice(0, splitIndex);
const testData = rows.slice(splitIndex);

// Remove labels from training and test data
// last column is label
const X_train = trainingData.map(row => row.slice(0, -1));
const X_test = testData.map(row => row.slice(0, -1));
const y_train = trainingData.map(row => row[row.length - 1]);  
const y_test = testData.map(row => row[row.length - 1]);

const knn = new KNNClassifier(25, 'euclidean', 'uniform');
knn.fit(X_train, y_train);

console.time("Prediction Time(Training)");
console.log("Training accuracy: ", knn.score(X_train, y_train));
console.timeEnd("Prediction Time(Training)");

console.time("Prediction Time(Testing)");
console.log("Test accuracy: ", knn.score(X_test, y_test));
console.timeEnd("Prediction Time(Testing)");

// y_pred = knn.predict(X_test);
// console.log("Predictions: ", X_test[0], y_pred[0], y_test[0]);
// console.log("Predictions: ", X_test[1], y_pred[1], y_test[1]);
// console.log("Predictions: ", X_test[2], y_pred[2], y_test[2]); 
// console.log("Predictions: ", X_test[3], y_pred[3], y_test[3]);
// console.log("Predictions: ", X_test[4], y_pred[4], y_test[4]);
// console.log("Predictions: ", X_test[5], y_pred[5], y_test[5]);
// console.log("Predictions: ", X_test[6], y_pred[6], y_test[6]);
// console.log("Predictions: ", X_test[7], y_pred[7], y_test[7]);
// console.log("Predictions: ", X_test[8], y_pred[8], y_test[8]);
// console.log("Predictions: ", X_test[9], y_pred[9], y_test[9]);
